java/ql/src/experimental/Security/CWE/CWE-552/UnsafeUrlForward.ql

```diff
@@ -14,20 +14,154 @@ import java
+        exists(MethodAccess ma2 |
+          isPathTraversalCheck(ma2) and
+          ma.getQualifier().(VarAccess).getVariable().getAnAccess() = ma2.getQualifier()
+        )
+        or
+        isFilePathMatch(ma)
+      ) and
+      (
+        not checkNoPathMatch(ma)
+        or
+        // non-match check needs decoding e.g. !path.startsWith("/WEB-INF/") won't detect /%57EB-INF/web.xml, which will be decoded and served by RequestDispatcher
+        checkNoPathMatch(ma) and
+        exists(MethodAccess ma2 |
+          isPathDecoding(ma2) and
+          ma.getQualifier().(VarAccess).getVariable().getAnAccess() = ma2.getQualifier()
+        )
+      ) and
+      this.asExpr() = ma.getQualifier()
+    )
   }
 }
 
+/**
+ * Holds if `ma` is a method call to check string content, which means an input string is not
+ * blindly trusted and helps to reduce FPs.
+ */
+predicate checkStringContent(MethodAccess ma, Expr expr) {
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod()
+      .hasName([
+          "charAt", "contains", "equals", "equalsIgnoreCase", "getBytes", "getChars", "indexOf",
+          "lastIndexOf", "length", "matches", "regionMatches", "replace", "replaceAll",
+          "replaceFirst", "substring"
+        ]) and
+  expr = ma.getQualifier()
+  or
+  (
+    ma.getMethod().getDeclaringType() instanceof TypeStringBuffer or
+    ma.getMethod().getDeclaringType() instanceof TypeStringBuilder
+  ) and
+  expr = ma.getAnArgument()
+}
+
+private class StringOperationSanitizer extends DataFlow::Node {
+  StringOperationSanitizer() { exists(MethodAccess ma | checkStringContent(ma, this.asExpr())) }
+}
+
+/**
+ * Holds if `expr` is an expression returned from null or empty string check.
+ */
+predicate isNullOrEmptyCheck(Expr expr) {
+  exists(ConditionBlock cb, ReturnStmt rt |
+    cb.controls(rt.getBasicBlock(), true) and
+    (
+      cb.getCondition().(EQExpr).getAnOperand() instanceof NullLiteral // if (path == null)
+      or
+      // if (path.equals(""))
+      exists(MethodAccess ma |
+        cb.getCondition() = ma and
+        ma.getMethod().getDeclaringType() instanceof TypeString and
+        ma.getMethod().hasName("equals") and
+        ma.getArgument(0).(CompileTimeConstantExpr).getStringValue() = ""
+      )
+    ) and
+    expr.getParent+() = rt
+  )
+}
+
+private class NullOrEmptyCheckSanitizer extends DataFlow::Node {
+  NullOrEmptyCheckSanitizer() { isNullOrEmptyCheck(this.asExpr()) }
+}
+
+/** Holds if `ma` is a virtual method call of Map::get or Object::toString. */
+predicate isVirtualMethod(MethodAccess ma, Expr expr) {
+  ma.getMethod().getDeclaringType() instanceof TypeObject and
+  ma.getMethod().hasName("toString") and
+  (expr = ma or expr = ma.getQualifier())
+  or
+  (
+    ma.getMethod().getDeclaringType().getASupertype*().hasQualifiedName("java.util", "Map") and
+    ma.getMethod().hasName(["get", "getOrDefault"])
+  ) and
+  (expr = ma or expr = ma.getAnArgument())
```

Comment:
Also, I understand this ad-hoc sanitizer was added because there are many FPs in certain projects that have classes implementing `Object.toString` and `Map.get[OrDefault]`. What I'm worried about is that we could lose potentially interesting flows that go through a `Map` because of this.

After reviewing the FPs, for instance in `apache/struts`, I'm seeing that the spurious flow is actually caused by the lack of call context in sources, e.g.

```java
public String toString() {
  return getTitle() + ":" + getSomethingElse(); // getTitle() is a source, and taints any call to `toString` from anywhere.
}
```

So I think this could be fixed using flow features instead. Try adding this predicate to your `Configuration` and remove `VirtualMethodSanitizer`:

```codeql
override DataFlow::FlowFeature getAFeature() {
  result instanceof DataFlow::FeatureHasSourceCallContext
}
```

That should get rid of FPs caused by sources lacking call context.

Reply:
Thanks @atorralba for reviewing this PR. The new approach works like a charm and helps to get rid of FPs caused by sources lacking call context.